node/bindnode/node.go

```diff
@@ -0,0 +1,1099 @@
+package bindnode
+
+import (
+	"fmt"
+	"reflect"
+	"strings"
+
+	ipld "github.com/ipld/go-ipld-prime"
+	basicnode "github.com/ipld/go-ipld-prime/node/basic"
+	"github.com/ipld/go-ipld-prime/schema"
+)
+
+// WrapNoSchema implements an ipld.Node given a pointer to a Go value.
+//
+// Same rules as PrototypeNoSchema apply.
+func WrapNoSchema(ptr interface{}) ipld.Node {
+	panic("TODO")
+	// ptrVal := reflect.ValueOf(ptr)
+	// if ptrVal.Kind() != reflect.Ptr {
+	// 	panic("must be a pointer")
+	// }
+	// return &_node{val: ptrVal.Elem()}
+}
+
+// Unwrap takes an ipld.Node implemented by one of the Wrap* APIs, and returns
+// a pointer to the inner Go value.
+//
+// Unwrap returns the input node if the node isn't implemented by this package.
+func Unwrap(node ipld.Node) (ptr interface{}) {
+	var val reflect.Value
+	switch node := node.(type) {
+	case *_node:
+		val = node.val
+	case *_nodeRepr:
+		val = node.val
+	default:
+		return node
+	}
+	if val.Kind() == reflect.Ptr {
+		panic("didn't expect val to be a pointer")
+	}
+	if !val.CanAddr() {
+		// Not addressable? Just return the interface as-is.
+		// TODO: This happens in some tests, figure out why.
+		return val.Interface()
+	}
+	return val.Addr().Interface()
+}
+
+// PrototypeNoSchema implements an ipld.NodePrototype given a Go pointer type.
+//
+// In this form, no IPLD schema is used; it is entirely inferred from the Go
+// type.
+//
+// Go types map to schema types in simple ways: Go string to schema String, Go
+// []byte to schema Bytes, Go struct to schema Map, and so on.
+//
+// A Go struct field is optional when its type is a pointer. Nullable fields are
+// not supported in this mode.
+func PrototypeNoSchema(ptrType interface{}) ipld.NodePrototype {
+	panic("TODO")
+	// typ := reflect.TypeOf(ptrType)
+	// if typ.Kind() != reflect.Ptr {
+	// 	panic("must be a pointer")
+	// }
+	// return &_prototype{goType: typ.Elem()}
+}
+
+// PrototypeOnlySchema implements an ipld.NodePrototype given an IPLD schema type.
+//
+// In this form, Go values are constructed with types inferred from the IPLD
+// schema, like a reverse of PrototypeNoSchema.
+func PrototypeOnlySchema(schemaType schema.Type) ipld.NodePrototype {
```

Comment:
This is *wild*.  :D  (You already blew my mind a couple days ago when telling me `reflect.StructOf` exists, you probably remember, heh.)

I have some rambling questions as I try to fit my head around when one would use this:

Do we know if this is going to be able to work for all schema typekinds and all their representations?  If there are any typekinds or representation strategies it's *not* going to be able to support, that could be a serious limitation of this entire feature (... possibly enough to make it questionable whether the feature is worth supporting?).

There are going to be some pretty interesting caveats on the usability of this too, right?  Like if it returns a struct... it's going to be some kind of slightly odd anonymous type, which can really only be usefully interacted with via more reflection or our Node interfaces?  ... Okay, I guess that may still sometimes be pretty useful, because it lets one get nice compact native memory layouts without having to use codegen.  That's cool.  But this feature wouldn't really be able to provide a "developer experience" win, right?

Reply:
> Do we know if this is going to be able to work for all schema typekinds and all their representations?

I'm pretty sure it should support all schema types, and by extension all representations too - representations are just some logic atop the same inferred Go types.

The big caveat is "what if I want my Go type to be slightly different", such as having your own Go type with methods, or using different types like `int` instead of `int64` for a field if you're absolutely sure the values will be small. That's where the answer will be "bring your own Go type and use the other API".

Reply:
Regarding anonymous types: it certainly has some disadvantages, like what I mentioned above. But the advantage is... it just works, without extra effort. No codegen, no having to declare the types yourself. If you have schema X and you want to convert data encoded in dagjson to dagcbor following the schema, you can do all of that with just the schema and PrototypeOnlySchema. The Go value you will hold in between will use anonymous types built at run-time, but for many simple use-cases, that's fine.

Reply:
But yes, for most real use cases like go-codec-dagpb where the type should be easy to use (methods, type switches, better error messages, etc), I do expect the code to provide both the schema and a Go type maintained by a human.